File: tests/release_workflows/derived-state-freshness/binding_workflow.py
"""Same-SHA native Python evidence for derived-state freshness."""

from __future__ import annotations

import argparse
import hashlib
import importlib.metadata
import json
from pathlib import Path
import uuid

import graphforge
import graphforge._graphforge_rs as native


def states(receipts: list[dict[str, object]]) -> list[str]:
    return [str(receipt["state"]) for receipt in receipts]


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--project", type=Path, required=True)
    parser.add_argument("--evidence", type=Path, required=True)
    parser.add_argument("--commit-sha", required=True)
    args = parser.parse_args()

    forge = graphforge.GraphForge(str(args.project))
    alice = forge.add_node("Person", name="Alice", summary="Graph systems")
    bob = forge.add_node("Person", name="Bob", summary="Native bindings")
    edge = forge.add_edge(alice, "KNOWS", bob)

    text_current = forge.index("Person", properties=["name"], rebuild=True)
    adjacency_current = forge.index_adjacency()
    embedding_v1 = forge.publish_caller_embeddings(
        "semantic",
        [
            {"node": alice, "vector": [1.0, 0.0]},
            {"node": bob, "vector": [0.0, 1.0]},
        ],
        dimensions=2,
        contract_version="derived-state-v1",
        source_projection={"label": "Person", "recipe": "v1"},
    )
    embedding_initial = forge.inspect_embedding_space_freshness("semantic")

    carol = forge.add_node("Person", name="Carol", summary="Fresh state")
    text_stale = forge.inspect_text_index("Person", properties=["name"])
    forge.execute(
        "MATCH ()-[r:KNOWS]->() WHERE r.edge_uuid = $edge_uuid DELETE r",
        {"edge_uuid": uuid.UUID(edge.uuid)},
    )
    adjacency_stale = forge.inspect_adjacency()
    text_rebuilt = forge.index("Person", properties=["name"], rebuild=True)
    adjacency_rebuilt = forge.rebuild_adjacency()

    embedding_v2 = forge.publish_caller_embeddings(
        "semantic",
        [
            {"node": alice, "vector": [0.0, 1.0]},
            {"node": bob, "vector": [1.0, 0.0]},
            {"node": carol, "vector": [1.0, 0.0]},
        ],
        dimensions=2,
        contract_version="derived-state-v2",
        source_projection={"label": "Person", "recipe": "v2"},
        replace=True,
    )
    embedding_replaced = forge.inspect_embedding_space_freshness("semantic")

    assert states([text_current, text_stale, text_rebuilt]) == [
        "current",
        "stale",
        "current",
    ]
    assert states([adjacency_current, adjacency_stale, adjacency_rebuilt]) == [
        "current",
        "stale",
        "current",
    ]
    assert embedding_initial["state"] == embedding_replaced["state"] == "fresh"
    assert embedding_initial["compatibility_id"] == embedding_v1
    assert embedding_replaced["compatibility_id"] == embedding_v2
    assert embedding_initial["generation_id"] != embedding_replaced["generation_id"]

    authority = {
        "text": forge.inspect_text_index("Person", properties=["name"]),
        "adjacency": forge.inspect_adjacency(),
        "embedding": forge.inspect_embedding_space_freshness("semantic"),
    }
    forge.close()
    reopened = graphforge.GraphForge(str(args.project))
    reopened_authority = {
        "text": reopened.inspect_text_index("Person", properties=["name"]),
        "adjacency": reopened.inspect_adjacency(),
        "embedding": reopened.inspect_embedding_space_freshness("semantic"),
    }
    reopened.close()
    assert reopened_authority == authority

    module_path = Path(native.__file__).resolve()
    evidence = {
        "schema_version": 1,
        "scenario_id": "derived-state-freshness",
        "binding": "python",
        "commit_sha": args.commit_sha,
        "text_states": states([text_current, text_stale, text_rebuilt]),
        "adjacency_states": states([adjacency_current, adjacency_stale, adjacency_rebuilt]),
        "compatibility_ids": [embedding_v1, embedding_v2],
        "generation_ids": [
            embedding_initial["generation_id"],
            embedding_replaced["generation_id"],
        ],
        "embedding_states": [embedding_initial["state"], embedding_replaced["state"]],
        "reopen_equal": reopened_authority == authority,
        "package_version": importlib.metadata.version("graphforge"),
        "native_version": graphforge.version(),
        "native_module_path": str(module_path),
        "native_module_sha256": hashlib.sha256(module_path.read_bytes()).hexdigest(),
    }
    args.evidence.write_text(json.dumps(evidence, indent=2, sort_keys=True) + "\n")


if __name__ == "__main__":
    main()

File: tests/release_workflows/derived-state-freshness/binding_workflow.mjs
// Same-SHA native Node evidence for derived-state freshness.

import assert from "node:assert/strict";
import { createHash } from "node:crypto";
import { createRequire } from "node:module";
import { readFileSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { pathToFileURL } from "node:url";
import { isDeepStrictEqual } from "node:util";

const args = Object.fromEntries(
  process.argv.slice(2).reduce((pairs, value, index, values) => {
    if (value.startsWith("--")) pairs.push([value.slice(2), values[index + 1]]);
    return pairs;
  }, []),
);
for (const name of ["project", "evidence", "commit-sha", "module"]) {
  if (!args[name]) throw new Error(`missing --${name}`);
}

const modulePath = resolve(args.module);
const { GraphForge, version } = await import(pathToFileURL(modulePath));
const packageVersion = JSON.parse(
  readFileSync(resolve(dirname(modulePath), "package.json"), "utf8"),
).version;
const forge = new GraphForge(resolve(args.project));
const alice = forge.addNode("Person", {
  name: "Alice",
  summary: "Graph systems",
});
const bob = forge.addNode("Person", {
  name: "Bob",
  summary: "Native bindings",
});
const edge = forge.addEdge(alice, "KNOWS", bob);

const textCurrent = forge.index("Person", {
  properties: ["name"],
  rebuild: true,
});
const adjacencyCurrent = forge.indexAdjacency();
const embeddingV1 = forge.publishCallerEmbeddings("semantic", {
  rows: [
    { node: alice, vector: [1, 0] },
    { node: bob, vector: [0, 1] },
  ],
  dimensions: 2,
  contractVersion: "derived-state-v1",
  sourceProjection: { label: "Person", recipe: "v1" },
});
const embeddingInitial = forge.inspectEmbeddingSpaceFreshness("semantic");

const carol = forge.addNode("Person", {
  name: "Carol",
  summary: "Fresh state",
});
const textStale = forge.inspectTextIndex("Person", ["name"]);
forge.execute(
  "MATCH ()-[r:KNOWS]->() WHERE r.edge_uuid = $edge_uuid DELETE r",
  { edge_uuid: { $uuid: edge.uuid } },
);
const adjacencyStale = forge.inspectAdjacency();
const textRebuilt = forge.index("Person", {
  properties: ["name"],
  rebuild: true,
});
const adjacencyRebuilt = forge.rebuildAdjacency();
const embeddingV2 = forge.publishCallerEmbeddings("semantic", {
  rows: [
    { node: alice, vector: [0, 1] },
    { node: bob, vector: [1, 0] },
    { node: carol, vector: [1, 0] },
  ],
  dimensions: 2,
  contractVersion: "derived-state-v2",
  sourceProjection: { label: "Person", recipe: "v2" },
  replace: true,
});
const embeddingReplaced = forge.inspectEmbeddingSpaceFreshness("semantic");

const states = (receipts) => receipts.map(({ state }) => state);
assert.deepEqual(states([textCurrent, textStale, textRebuilt]), [
  "current",
  "stale",
  "current",
]);
assert.deepEqual(states([adjacencyCurrent, adjacencyStale, adjacencyRebuilt]), [
  "current",
  "stale",
  "current",
]);
assert.equal(embeddingInitial.state, "fresh");
assert.equal(embeddingReplaced.state, "fresh");
assert.equal(embeddingInitial.compatibilityId, embeddingV1);
assert.equal(embeddingReplaced.compatibilityId, embeddingV2);
assert.notEqual(embeddingInitial.generationId, embeddingReplaced.generationId);

const authority = {
  text: forge.inspectTextIndex("Person", ["name"]),
  adjacency: forge.inspectAdjacency(),
  embedding: forge.inspectEmbeddingSpaceFreshness("semantic"),
};
forge.close();
const reopened = new GraphForge(resolve(args.project));
const reopenedAuthority = {
  text: reopened.inspectTextIndex("Person", ["name"]),
  adjacency: reopened.inspectAdjacency(),
  embedding: reopened.inspectEmbeddingSpaceFreshness("semantic"),
};
reopened.close();
const reopenEqual = isDeepStrictEqual(reopenedAuthority, authority);
assert.deepEqual(reopenedAuthority, authority);

const require = createRequire(import.meta.url);
const addonPaths = Object.entries(require.cache)
  .filter(
    ([path, cached]) =>
      path.endsWith(".node") && cached.exports.GraphForge === GraphForge,
  )
  .map(([path]) => path);
assert.equal(addonPaths.length, 1, `loaded addons: ${addonPaths.join(", ")}`);
const addonPath = resolve(addonPaths[0]);
const addonSha256 = createHash("sha256")
  .update(readFileSync(addonPath))
  .digest("hex");
writeFileSync(
  resolve(args.evidence),
  `${JSON.stringify(
    {
      schema_version: 1,
      scenario_id: "derived-state-freshness",
      binding: "node",
      commit_sha: args["commit-sha"],
      package_version: packageVersion,
      text_states: states([textCurrent, textStale, textRebuilt]),
      adjacency_states: states([
        adjacencyCurrent,
        adjacencyStale,
        adjacencyRebuilt,
      ]),
      compatibility_ids: [embeddingV1, embeddingV2],
      generation_ids: [
        embeddingInitial.generationId,
        embeddingReplaced.generationId,
      ],
      embedding_states: [embeddingInitial.state, embeddingReplaced.state],
      reopen_equal: reopenEqual,
      native_version: version(),
      native_module_path: addonPath,
      native_module_sha256: addonSha256,
    },
    null,
    2,
  )}\n`,
);
